# (실습) 파일

파일 저장 디렉토리 지정

In [4]:
from pathlib import Path

data_path = Path() / "data"

지정된 디렉토리 생성

In [5]:
data_path.mkdir(parents=True, exist_ok=True)

파일 다운로드와 저장 함수

In [8]:
import urllib.request

def myWget(filename):
    base_url = "https://raw.githubusercontent.com/codingalzi/pybook/master/jupyter-book/data/"
    file_url = base_url + filename

    target_path = data_path / filename

    return urllib.request.urlretrieve(file_url, target_path)

**문제 1**

`shopA.txt` 파일은 쇼핑몰A에서 판매하는 상품의 가격을 담고 있음을 확인해보자.
먼저 해당 파일을 다운로드 한다.

In [9]:
myWget("shopA.txt")

(PosixPath('data/shopA.txt'), <http.client.HTTPMessage at 0x7f658059f2e0>)

이제 파일 전체 내용을 출력하는 코드를 작성하라.

힌트: `open()` 함수와 `with-as` 명령문 활용

In [ ]:
# 코드를 작성하세요.



**문제 2**

`shopA.txt` 파일의 내용을 확인하면, 오타가 있다. 
'오레ㄴ지' 를 '오렌지'로 변경한 후에
`shopA.txt` 파일을 열어 오타가 수정되었는지를 확인하여라.  

힌트: 파일의 `read()` 메서드, 문자열의 `replace()` 메서드

* 파일 읽기: `read()` 메서드 활용

In [10]:
# 파일내용을 하나의 문자열로 생성하는 코드를 작성하라.



* 오타 수정: `replace()` 문자열 메서드 활용

In [11]:
# 오타를 수정하는 코드를 작성하라.



* 파일 저장

In [12]:
# 오타가 수정된 문자열을 파일로 저장하는 코드를 작성하라.



* 파일 내용 확인

In [13]:
# 오타가 수정되었음을 확인하는 코드를 작성하라.



**문제 3**

상품명과 가격을 키-값의 쌍으로 갖는 아래 모양의 딕셔너리를 만들어라.
단, 오타가 수정된 파일을 이용해야 한다.

```python
{'우유': 2540,
 '계란': 7480,
 '생수': 980,
 '짜장라면': 3220,
 '두부': 1450,
 '콩나물': 1680,
 '김': 5480,
 '닭고기': 5980,
 '식빵': 2480,
 '바나나': 4980,
 '오렌지': 990,
 '카레': 2480,
 '만두': 6980,
 '어묵': 7980,
 '참치': 11880,
 '김치': 7980,
 '간장': 10800}
```

## 4단계

4단계) 상품명과 가격을 담고 있는 파일의 이름을 입력받아 상품명을 key, 가격을 value로 하는 딕셔너리를 리턴하는 함수`shopping`를 만들어라. 

예를 들면, 아래와 같다(딕셔너리의 항목 일부 생략함)
```
In [5]: shopping('/content/shopA.txt')
Out[5]: {'우유' : 2540, '계란' : 7480, '생수' : 980}
```

[guide] 3단계까지 작성한 코드를 이용하여 함수를 정의하자

In [ ]:
# 함수 정의

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue

  shop.close()

  return shop_dict



## 함수를 이용하여 리턴된 딕셔너리를 확인하자
shop_path = "drive/MyDrive/shopA_revised.txt"
shopA_dict = shopping(shop_path)
print(shopA_dict)

{'우유': 2540, '계란': 7480, '생수': 980, '짜장라면': 3220, '두부': 1450, '콩나물': 1680, '김': 5480, '닭고기': 5980, '식빵': 2480, '바나나': 4980, '오렌지': 990, '카레': 2480, '만두': 6980, '어묵': 7980, '참치': 11880, '김치': 7980, '간장': 10800}


## 5단계

5단계) 판매하는 상품을 보여준 다음, 사용자가 상품을 입력하면 상품의 가격이 얼마인지를 알려주는 코드를 작성하여라.

[Guide] 

1. 판매하는 상품을 보여준다.

2. 4단계의 함수를 이용하여 상품과 가격을 Key, Value로 담은 딕셔너리를 return 받는다.

3. 상품을 입력받는다.

4. 2.에서 구한 딕셔너리로 가격을 보여준다  


In [ ]:
# 함수 정의 

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue
  
  shop.close()

  return shop_dict


# 1. 상품 보여주기
shop_path = "drive/MyDrive/shopA_revised.txt"

with open(shop_path, 'r') as shop :
  print(shop.read())

# 2. 딕셔너리 리턴
shopA_dict = shopping(shop_path)

# 3. 입력받기
try:
  want_to_buy = input("구매하실 상품을 입력해주세요 : ")
  print( want_to_buy + "의 가격은 " + str(shopA_dict[want_to_buy]) + "원입니다.") # 쇼핑몰에 없는 상품을 입력하면 KeyError가 발생한다. -> 에러처리
except KeyError:
  print("상품명을 확인해주세요.") 

#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원
구매하실 상품을 입력해주세요 : 김치
김치의 가격은 7980원입니다.


## 6단계

6단계) `shopB.txt`파일은 쇼핑몰B에서 판매하는 상품의 가격을 담고 있다.  
사용자가 상품을 입력하면, 쇼핑몰A,B 중 어느 쇼핑몰에서 구입하는 것이 얼마나 저렴한지를 보여주는 코드를 작성하여라.  

[Guide] 

1. shopping 함수를 이용하여 쇼핑몰 A와 쇼핑몰 B의 딕셔너리를 구한 뒤,
2. 구매를 원하는 상품을 입력받고
3. 가격을 비교하자

In [ ]:
# 구글 드라이브를 연동하지 않은 경우에는 shopB.txt 를 따로 올려주어야 한다.
# 여기선 1단계에서 연동한 구글 드라이브를 활용한다.

In [ ]:
# 함수 정의 

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue

  shop.close()

  return shop_dict

def show_items(shop_path):
  with open(shop_path, 'r') as shop :
   print(shop.read())

# 상품 보여주기
shopA_path = "drive/MyDrive/shopA_revised.txt"
shopB_path = "drive/MyDrive/shopB.txt"

show_items(shopA_path)
print("\n") # 이쁘게 print하기 위해 추가.
show_items(shopB_path)

# 딕셔너리 리턴
shopA_dict = shopping(shopA_path)
shopB_dict = shopping(shopB_path)

try :
  
  # 입력받기
  want_to_buy = input("구매하실 상품을 입력해주세요 : ")

  # 가격비교하기
  if   shopA_dict[want_to_buy] > shopB_dict[want_to_buy] : 
    dif = shopA_dict[want_to_buy] - shopB_dict[want_to_buy]
    print("ShopB에서 " + str(dif) + "원 더 저렴합니다.")

  elif shopA_dict[want_to_buy] < shopB_dict[want_to_buy] : 
    dif = shopB_dict[want_to_buy] - shopA_dict[want_to_buy]
    print("ShopA에서 " + str(dif) + "원 더 저렴합니다.")

  else :
    print("가격이 동일합니다.")

except KeyError :
  print("상품명을 확인해주세요.")  


#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원


#쇼핑몰 B

우유 2270원
계란 7520원
생수 1010원
짜장라면 3210원
두부 1150원
콩나물 1980원
김 5290원
닭고기 6000원
식빵 1990원
바나나 4990원
오렌지 1050원
카레 2400원
만두 7400원
어묵 6980원
참치 11580원
김치 8480원
간장 9900원
구매하실 상품을 입력해주세요 : 참치
ShopB에서 300원 더 저렴합니다.


```








.
```

연습을 더 하고 싶다면, 아래의 문제를 풀어보세요!  

**[추가문제]**

## 7단계

7단계) 사용자가 프로그램이 종료되길 원할 때까지 상품을 입력받아라.

[Guide] while문을 이용하여서 작성할 수 있다.

1. continue_shopping = 1 을 할당하고, 특정 조건에서 continue_shopping = 0이 되게 바꾼다.

2. 특정 조건은 사용자가 프로그램 종료를 원할 때이다.

In [ ]:
# 함수 정의

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue

  shop.close()

  return shop_dict

def show_items(shop_path):
  with open(shop_path, 'r') as shop :
   print(shop.read())

#------------------------------------------------------------#

# 쇼핑몰 보여주기
shopA_path = "drive/MyDrive/shopA_revised.txt"
shopB_path = "drive/MyDrive/shopB.txt"
show_items(shopA_path)
print("\n")
show_items(shopB_path)

# 딕셔너리 리턴
shopA_dict = shopping(shopA_path)
shopB_dict = shopping(shopB_path)

# 쇼핑을 계속할지 알려주는 변수
continue_shopping = 1

while continue_shopping :


  try:
    want_to_buy = input("구매하실 상품을 입력해주세요 : ")

    # 가격비교
    if shopA_dict[want_to_buy] > shopB_dict[want_to_buy] : 
      dif = shopA_dict[want_to_buy] - shopB_dict[want_to_buy]
      print("ShopB에서 " + str(dif) + "원 더 저렴합니다.")

    elif shopA_dict[want_to_buy] < shopB_dict[want_to_buy] : 
      dif = shopB_dict[want_to_buy] - shopA_dict[want_to_buy]
      print("ShopA에서 " + str(dif) + "원 더 저렴합니다.")

    else :
      print("가격이 동일합니다.")

  except KeyError :
    print("상품명을 확인해주세요.")

  want_to_continue = input("계속 쇼핑하시겠습니까? (y/n)")
  if want_to_continue == 'n':
    continue_shopping = 0
    print("쇼핑을 종료합니다 :)")

#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원


#쇼핑몰 B

우유 2270원
계란 7520원
생수 1010원
짜장라면 3210원
두부 1150원
콩나물 1980원
김 5290원
닭고기 6000원
식빵 1990원
바나나 4990원
오렌지 1050원
카레 2400원
만두 7400원
어묵 6980원
참치 11580원
김치 8480원
간장 9900원
구매하실 상품을 입력해주세요 : 김치
ShopA에서 500원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)y
구매하실 상품을 입력해주세요 : 간장
ShopB에서 900원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)n
쇼핑을 종료합니다 :)


## 8단계

8단계) 아래의 내용을 담은 `shopping_record.txt` 파일을 만들어라. 
 
* 사용자가 입력한 상품명.
* 상품의 가격. 이때 두 쇼핑몰의 상품 가격 중 저렴한 것으로 저장. 
* 쇼핑몰 이름.

예를 들어, 사용자가 생수(쇼핑몰A 980원, 쇼핑몰B 1010원)를 입력했다면, 아래와 같이 저장한다. 

```
생수 980원 쇼핑몰A
```

* 주문한 상품의 총 가격.

[Guide]

1. 사용자가 입력한 상품명, 가장 저렴한 가격, 저렴한 쇼핑몰을 기록하는 리스트를 할당하자.

2. 계속해서 리스트에 append해주고 

3. 프로그램이 종료된 후 리스트를 이용하여서 txt 파일을 쓰자



In [ ]:
# 함수 정의 

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원",""))
    except:
      continue

  shop.close()

  return shop_dict

def show_items(shop_path):
  with open(shop_path, 'r') as shop :
   print(shop.read())


def list_append(shop_record, want_to_buy, price, shop):

  """ 매개변수 순서: 기록할 리스트, 상품명, 가격, 쇼핑몰"""

  # 입력받은 상품 append
  shop_record.append(want_to_buy)
  # 더 저렴한 가격 append라인에 붙이기
  shop_record.append(price)
  # 더 저렴한 쇼핑몰 append라인에 붙이기
  shop_record.append(shop)

#------------------------------------------------------------#


# 쇼핑몰 경로
shopA_path = "drive/MyDrive/shopA_revised.txt"
shopB_path = "drive/MyDrive/shopB.txt"

# 쇼핑몰 보여주기
show_items(shopA_path)
print("\n")
show_items(shopB_path)

# 딕셔너리 리턴
shopA_dict = shopping(shopA_path)
shopB_dict = shopping(shopB_path)

# shopping_record.txt 생성.
shopping_record = open("drive/MyDrive/shopping_record.txt",'w') # 'w'을 매개변수로 주어 write 모드로 open

# 상품명과 저렴한 가격, 쇼핑몰을 기록할 리스트
shopping_record_content = []

# 쇼핑을 계속할지 알려주는 변수
continue_shopping = 1

while continue_shopping :

  # 입력받기

  try :
    want_to_buy = input("구매하실 상품을 입력해주세요 : ")


    if shopA_dict[want_to_buy] > shopB_dict[want_to_buy] : 
      dif = shopA_dict[want_to_buy] - shopB_dict[want_to_buy]
      print("ShopB에서 " + str(dif) + "원 더 저렴합니다.")
        
      # shopping_record_content에 append
      # 리스트, 상품, 가격,  쇼핑몰 순
      list_append(shopping_record_content, want_to_buy, shopB_dict[want_to_buy], "쇼핑몰B")

    elif shopA_dict[want_to_buy] < shopB_dict[want_to_buy] : 
      dif = shopB_dict[want_to_buy] - shopA_dict[want_to_buy]
      print("ShopA에서 " + str(dif) + "원 더 저렴합니다.")

      # shopping_record_content에 append
      # 리스트, 상품, 가격,  쇼핑몰 순
      list_append(shopping_record_content, want_to_buy, shopA_dict[want_to_buy],"쇼핑몰A")


    else :
      print("가격이 동일합니다.")
      # 가격이 동일하면 쇼핑몰 A 선택
      # shopping_record_content에 append
      # 리스트, 상품, 가격,  쇼핑몰 순
      list_append(shopping_record_content, want_to_buy, shopA_dict[want_to_buy], "쇼핑몰A")

  except : 
    print("상품명을 확인해주세요.")
    continue


  want_to_continue = input("계속 쇼핑하시겠습니까? (y/n)")
  if want_to_continue == 'n':
    continue_shopping = 0
    print("쇼핑을 종료합니다 :)")

print("\n") # 이쁘게 print하기 위해서 추가

shopping_record_text = "### 구매내역 ###\n"
for i in range(len(shopping_record_content) // 3):
  shopping_record_text += shopping_record_content[i*3] + " "
  shopping_record_text += str(shopping_record_content[i*3 + 1]) + "원 "
  shopping_record_text += shopping_record_content[i*3 + 2] + "\n"  # 줄 바꿈을 위해서 \n을 추가

shopping_record.write(shopping_record_text) # shopping_record에 shopping_record_text를 써주기 
shopping_record.close() # open한 파일은 close를 해준다

# shopping_record를 읽어서 확인해보기.

with open("drive/MyDrive/shopping_record.txt",'r') as shopping_record:
  print(shopping_record.read())

#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원


#쇼핑몰 B

우유 2270원
계란 7520원
생수 1010원
짜장라면 3210원
두부 1150원
콩나물 1980원
김 5290원
닭고기 6000원
식빵 1990원
바나나 4990원
오렌지 1050원
카레 2400원
만두 7400원
어묵 6980원
참치 11580원
김치 8480원
간장 9900원
구매하실 상품을 입력해주세요 : 김치
ShopA에서 500원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)y
구매하실 상품을 입력해주세요 : 바나나
ShopA에서 10원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)y
구매하실 상품을 입력해주세요 : 식빵
ShopB에서 490원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)n
쇼핑을 종료합니다 :)


### 구매내역 ###
김치 7980원 쇼핑몰A
바나나 4980원 쇼핑몰A
식빵 1990원 쇼핑몰B



## 9단계

9단계) 사용자에게 구매할 상품의 개수도 입력받아 `shopping_record.txt` 파일에 저장하여라.  

예를 들어, 저장된 내용은 아래와 같다. 
```
생수 980원 2개 쇼핑몰A
우유 2270원 1개 쇼핑몰B
오렌지 990원 2개 쇼핑몰A
=======================
총 가격 : 6210원
```

[Guide] 

1. 상품 개수도 입력을 같이받자
2. 상품명, 가장 저렴한 가격, 구매수량, 쇼핑몰을 기록하는 리스트를 만들자
3. 프로그램이 종료되면, 2.의 리스트를 이용하여서 총 가격과 txt파일에 요구되는 방식대로 기록하자

In [ ]:
# 함수 정의 

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue

  shop.close()

  return shop_dict

def show_items(shop_path):
  with open(shop_path, 'r') as shop :
   print(shop.read())


def list_append_9step(shopping_record_content, want_to_buy, price, quantity, shop):

  """ 매개변수 순서: 기록할 리스트, 상품명, 가격, 쇼핑몰"""

  # 입력받은 상품 append
  shopping_record_content.append(want_to_buy)
  # 더 저렴한 가격 append라인에 붙이기
  shopping_record_content.append(price)
  # 입력받은 수량 append
  shopping_record_content.append(quantity)
  # 더 저렴한 쇼핑몰 append라인에 붙이기
  shopping_record_content.append(shop)

#-----------------------------------------------------------------------------#


# 쇼핑몰 경로
shopA_path = "drive/MyDrive/shopA_revised.txt"
shopB_path = "drive/MyDrive/shopB.txt"

# 쇼핑몰 보여주기
show_items(shopA_path)
print("\n")
show_items(shopB_path)

# 딕셔너리 리턴
shopA_dict = shopping(shopA_path)
shopB_dict = shopping(shopB_path)

# shopping_record 생성.
shopping_record = open("drive/MyDrive/shopping_record.txt",'w')
shopping_record_content = []

# 쇼핑을 계속할지 알려주는 변수
continue_shopping = 1

while continue_shopping :


  try :
    # 상품명 입력받기
    want_to_buy = input("구매하실 상품을 입력해주세요 : ")

    # 수량 입력받기
    try : 
      how_many = int(input("구매하실 상품의 수량을 입력해주세요 : "))

      # 음수이거나 0을 입력하는 경우 처리
      if how_many <= 0 :
        print("수량을 정확히 입력해주세요")
        continue

    # 정수를 입력하지 않은 경우 예외처리
    except ValueError :
      print("수량을 정확히 입력해주세요")

    # 가격비교
    if shopA_dict[want_to_buy] > shopB_dict[want_to_buy] : 
      dif = shopA_dict[want_to_buy] - shopB_dict[want_to_buy]
      print("ShopB에서 " + str(dif) + "원 더 저렴합니다.")
        
      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopB_dict[want_to_buy], how_many, "쇼핑몰B")

    elif shopA_dict[want_to_buy] < shopB_dict[want_to_buy] : 
      dif = shopB_dict[want_to_buy] - shopA_dict[want_to_buy]
      print("ShopA에서 " + str(dif) + "원 더 저렴합니다.")

      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopA_dict[want_to_buy], how_many, "쇼핑몰A")


    else :
      print("가격이 동일합니다.")
      # 가격이 동일하면 쇼핑몰 A 선택
      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopA_dict[want_to_buy], how_many, "쇼핑몰A")

  except : 
    print("상품명을 확인해주세요.")
    continue

  want_to_continue = input("계속 쇼핑하시겠습니까? (y/n)")
  if want_to_continue == 'n':
    continue_shopping = 0
    print("쇼핑을 종료합니다 :)")

shopping_record_text = "### 구매내역 ###\n"

# 총 가격 계산
price_to_pay = 0

for i in range(len(shopping_record_content) // 4):
  shopping_record_text += shopping_record_content[i*4] + " "
  shopping_record_text += str(shopping_record_content[i*4 + 1]) + "원 "
  shopping_record_text += str(shopping_record_content[i*4 + 2]) + "개 "
  shopping_record_text += shopping_record_content[i*4 + 3] + "\n" # 줄 바꿈을 위해서 \n을 추가
  
  price_to_pay += shopping_record_content[i*4 + 1] * shopping_record_content[i*4 + 2] # 가격 x 수량

shopping_record_text += "=======================\n"
shopping_record_text += "총 가격 : " + str(price_to_pay) + "원" 

shopping_record.write(shopping_record_text) # shopping_record에 text를 쓰자.
shopping_record.close() # open한 파일은 close해준다

# shopping_record를 읽어서 확인해보기.

with open("drive/MyDrive/shopping_record.txt",'r') as shopping_record:
  print(shopping_record.read())

#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원


#쇼핑몰 B

우유 2270원
계란 7520원
생수 1010원
짜장라면 3210원
두부 1150원
콩나물 1980원
김 5290원
닭고기 6000원
식빵 1990원
바나나 4990원
오렌지 1050원
카레 2400원
만두 7400원
어묵 6980원
참치 11580원
김치 8480원
간장 9900원
구매하실 상품을 입력해주세요 : 만두
구매하실 상품의 수량을 입력해주세요 : 2
ShopA에서 420원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)카레
구매하실 상품을 입력해주세요 : 카레
구매하실 상품의 수량을 입력해주세요 : -100
수량을 정확히 입력해주세요
구매하실 상품을 입력해주세요 : 카레
구매하실 상품의 수량을 입력해주세요 : 2
ShopB에서 80원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)y
구매하실 상품을 입력해주세요 : 우유
구매하실 상품의 수량을 입력해주세요 : 10
ShopB에서 270원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)n
쇼핑을 종료합니다 :)
### 구매내역 ###
만두 6980원 2개 쇼핑몰A
카레 2400원 2개 쇼핑몰B
우유 2270원 10개 쇼핑몰B
총 가격 : 41460원


## 10단계

10단계) 쇼핑몰별로 정리하여 `shopping_record.txt` 파일에 저장하여라.  
예를 들어, 저장된 내용은 아래와 같다. 

```
생수 980원 2개 쇼핑몰A
오렌지 990원 2개 쇼핑몰A
우유 2270원 1개 쇼핑몰B
=======================
쇼핑몰A 가격 : 3940원
쇼핑몰B 가격 : 2270원
총 가격 : 6210원
```

[Guide]

9단계와 거의 유사하다.

대신, 총 가격을 계산할 때, 쇼핑몰 A와 쇼핑몰 B를 나누어 계산해주고, 총 가격은 쇼핑몰A 가격과 쇼핑몰B 가격을 합하여 구해주자

In [ ]:
# 함수 정의 

def shopping(shop_path) :
  shop = open(shop_path,'r')

  shop_dict = {}

  for line in shop:
    try : 
      (product, price) = line.strip().split() 
      shop_dict[product] = int(price.replace("원","")) 
    except:
      continue

  shop.close()

  return shop_dict

def show_items(shop_path):
  with open(shop_path, 'r') as shop :
   print(shop.read())


def list_append_9step(shopping_record_content, want_to_buy, price, quantity, shop):

  """ 매개변수 순서: 기록할 리스트, 상품명, 가격, 쇼핑몰"""

  # 입력받은 상품 append
  shopping_record_content.append(want_to_buy)
  # 더 저렴한 가격 append라인에 붙이기
  shopping_record_content.append(price)
  # 입력받은 수량 append
  shopping_record_content.append(quantity)
  # 더 저렴한 쇼핑몰 append라인에 붙이기
  shopping_record_content.append(shop)

#-----------------------------------------------------------------------------#


# 쇼핑몰 경로
shopA_path = "drive/MyDrive/shopA_revised.txt"
shopB_path = "drive/MyDrive/shopB.txt"

# 쇼핑몰 보여주기
show_items(shopA_path)
print("\n")
show_items(shopB_path)

# 딕셔너리 리턴
shopA_dict = shopping(shopA_path)
shopB_dict = shopping(shopB_path)

# shopping_record 생성.
shopping_record = open("drive/MyDrive/shopping_record.txt",'w')
shopping_record_content = []

# 쇼핑을 계속할지 알려주는 변수
continue_shopping = 1

while continue_shopping :


  try :
    # 상품명 입력받기
    want_to_buy = input("구매하실 상품을 입력해주세요 : ")

    # 수량 입력받기
    try : 
      how_many = int(input("구매하실 상품의 수량을 입력해주세요 : "))

      # 음수이거나 0을 입력하는 경우 처리
      if how_many <= 0 :
        print("수량을 정확히 입력해주세요")
        continue

    # 정수를 입력하지 않은 경우 예외처리
    except ValueError :
      print("수량을 정확히 입력해주세요")

    # 가격비교
    if shopA_dict[want_to_buy] > shopB_dict[want_to_buy] : 
      dif = shopA_dict[want_to_buy] - shopB_dict[want_to_buy]
      print("ShopB에서 " + str(dif) + "원 더 저렴합니다.")
        
      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopB_dict[want_to_buy], how_many, "쇼핑몰B")

    elif shopA_dict[want_to_buy] < shopB_dict[want_to_buy] : 
      dif = shopB_dict[want_to_buy] - shopA_dict[want_to_buy]
      print("ShopA에서 " + str(dif) + "원 더 저렴합니다.")

      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopA_dict[want_to_buy], how_many, "쇼핑몰A")


    else :
      print("가격이 동일합니다.")
      # 가격이 동일하면 쇼핑몰 A 선택
      # shopping_record_content에 append
      # 리스트, 상품, 가격, 수량, 쇼핑몰 순
      list_append_9step(shopping_record_content, want_to_buy, shopA_dict[want_to_buy], how_many, "쇼핑몰A")

  except : 
    print("상품명을 확인해주세요.")
    continue

  want_to_continue = input("계속 쇼핑하시겠습니까? (y/n)")
  if want_to_continue == 'n':
    continue_shopping = 0
    print("쇼핑을 종료합니다 :)")

shopping_record_text = "### 구매내역 ###\n"
# 총 가격 계산
price_to_pay_shopA = 0
price_to_pay_shopB = 0

for i in range(len(shopping_record_content) // 4):
  shopping_record_text += shopping_record_content[i*4] + " "
  shopping_record_text += str(shopping_record_content[i*4 + 1]) + "원 "
  shopping_record_text += str(shopping_record_content[i*4 + 2]) + "개 "
  shopping_record_text += shopping_record_content[i*4 + 3] + "\n"
  
  ## 쇼핑몰A와 쇼핑몰B를 구분하여 계산
  if shopping_record_content[i*4 + 3] == "쇼핑몰A" :
    price_to_pay_shopA += shopping_record_content[i*4 + 1] * shopping_record_content[i*4 + 2]
  else : 
    price_to_pay_shopB += shopping_record_content[i*4 + 1] * shopping_record_content[i*4 + 2]

  ## 총 가격은 합쳐서 계산
  price_to_pay = price_to_pay_shopA  +  price_to_pay_shopB

shopping_record_text += "=======================\n"
shopping_record_text += "쇼핑몰A 가격 : " + str(price_to_pay_shopA) + "원\n" 
shopping_record_text += "쇼핑몰B 가격 : " + str(price_to_pay_shopB) + "원\n" 
shopping_record_text += "총 가격 : " + str(price_to_pay) + "원" 

shopping_record.write(shopping_record_text)
shopping_record.close()

# shopping_record를 읽어서 확인해보기.

with open("drive/MyDrive/shopping_record.txt",'r') as shopping_record:
  print(shopping_record.read())

#쇼핑몰 A

우유 2540원
계란 7480원
생수 980원
짜장라면 3220원
두부 1450원
콩나물 1680원
김 5480원
닭고기 5980원
식빵 2480원
바나나 4980원
오렌지 990원
카레 2480원
만두 6980원
어묵 7980원
참치 11880원
김치 7980원
간장 10800원


#쇼핑몰 B

우유 2270원
계란 7520원
생수 1010원
짜장라면 3210원
두부 1150원
콩나물 1980원
김 5290원
닭고기 6000원
식빵 1990원
바나나 4990원
오렌지 1050원
카레 2400원
만두 7400원
어묵 6980원
참치 11580원
김치 8480원
간장 9900원
구매하실 상품을 입력해주세요 : 어묵
구매하실 상품의 수량을 입력해주세요 : 10
ShopB에서 1000원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)y
구매하실 상품을 입력해주세요 : 바나나
구매하실 상품의 수량을 입력해주세요 : 2
ShopA에서 10원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)계란
구매하실 상품을 입력해주세요 : 1
구매하실 상품의 수량을 입력해주세요 : 1
상품명을 확인해주세요.
구매하실 상품을 입력해주세요 : 계란
구매하실 상품의 수량을 입력해주세요 : 1
ShopA에서 40원 더 저렴합니다.
계속 쇼핑하시겠습니까? (y/n)n
쇼핑을 종료합니다 :)
### 구매내역 ###
어묵 6980원 10개 쇼핑몰B
바나나 4980원 2개 쇼핑몰A
계란 7480원 1개 쇼핑몰A
쇼핑몰A 가격 : 17440원
쇼핑몰B 가격 : 69800원
총 가격 : 87240원


**수고하셨습니다 :)**